### ISSUE TITLE
Cherry pick multiple bug fixes to r0.5

### ISSUE THREAD

**Issue by fpj** [pravega/pravega#3719]
**Problem description**
I'll update this issue with a list of merges I'll be putting together in a branch to be cherry-picked to `r0.5`.

**Problem location**
Release branch `0.5`.

**Suggestions for an improvement**
Cherry pick relevant merges.

**Comment by fpj** [comment]
The list of issues that I'm planning to incorporate:

```
Issue 3708: Sporadic failure of HostStoreTest.zkHostStoreTests (#3712)
Issue #3333, #2572: Keep completed segments in position objects. (#3364)
Issue 3673: New controller instance is not updating segment container map unless there is an update (#3682)
Issue 3649: Invoke grgit plugin only if git repository is present. (#3667)
Issue 3643: Fix reliability issues of standalone mode tests (#3656)
Issue 3659: Update the config.properties file to document hdfs.replaceDataNodesOnFailure usage. (#3680)
Issue 3659: Set hdfs.replaceDataNodesOnFailure = false in K8s system tests. (#3660)
```

**Comment by andreipaduroiu** [comment]
Let's add this too:

**Comment by andreipaduroiu** [comment]
And this? 
https://github.com/pravega/pravega/pull/3709

**Comment by tkaitchuck** [comment]
I think we should also include:
Issue 3632: Add session to connection before it is established (#3639)
and
Issue #3624: Fix backward compatibility issue with WireCommands. (#3640)

As these fix issues with connection pooling.

**Comment by fpj** [comment]
@tkaitchuck these are already in:

https://github.com/pravega/pravega/commits/r0.5

**Comment by tkaitchuck** [comment]
LGTM

**Comment by shrids** [comment]
@fpj  @tkaitchuck  I thinks https://github.com/pravega/pravega/pull/3654 this should be included too.

**Comment by fpj** [comment]
I have added the following four merges:

```
Issue 3638: Ensure the AsyncSegmentInputStreamImpl handles ConnectionFailedException during sendAsync. (#3654)
Issue 3702: Prevent suppression of stack traces in tests (#3703)
Issue 2830: PingTxnStatus now informs if the Transaction is closed (#3704)
Issue 3710: (SegmentStore) Delete Segment if Empty Bug Fix (#3711)
```

**Comment by RaulGracia** [comment]
Closed via #3720.
